File: tests/SnapBrief.App.Imaging.Tests/TextMarkMetricsTests.cs
using System.Windows;
using SnapBrief.App;

namespace SnapBrief.App.Imaging.Tests;

// The box of a caption is measured, not guessed: everything that finds a mark under the pointer
// works on these numbers, and they are the same numbers the export draws with.
public sealed class TextMarkMetricsTests
{
    [Fact]
    public void Measure_GivesAWideBoxForAWordAndAtLeastTheHeightOfTheLetters()
    {
        var size = TextMarkMetrics.Measure("Привет", 32);

        Assert.True(size.Width > size.Height, $"A word must be wider than it is tall, it measured {size.Width}x{size.Height}.");
        Assert.True(size.Height >= 32);
    }

    [Fact]
    public void Measure_ReadsCyrillicAndLatinAlike()
    {
        var cyrillic = TextMarkMetrics.Measure("Привет", 20);
        var latin = TextMarkMetrics.Measure("Hello", 20);

        Assert.True(cyrillic.Width > 0);
        Assert.True(latin.Width > 0);
        Assert.True(TextMarkMetrics.Measure("Привет, мир", 20).Width > cyrillic.Width);
    }

    [Fact]
    public void Measure_OfNothingIsStillABoxWithRoomForACaret()
    {
        var size = TextMarkMetrics.Measure(string.Empty, 20);

        Assert.True(size.Width > 0);
        Assert.True(size.Height >= 20);
    }

    [Fact]
    public void Measure_GrowsWithTheSizeOfTheLetters()
    {
        Assert.True(TextMarkMetrics.Measure("Привет", 48).Width > TextMarkMetrics.Measure("Привет", 16).Width);
    }

    [Theory]
    [InlineData(0, TextMarkMetrics.MinimumFontSize)]
    [InlineData(4, TextMarkMetrics.MinimumFontSize)]
    [InlineData(20, 20)]
    [InlineData(400, TextMarkMetrics.MaximumFontSize)]
    public void Clamp_BringsASizeNobodyCanDrawBackIntoRange(double stored, double expected)
    {
        Assert.Equal(expected, TextMarkMetrics.Clamp(stored));
    }

    [Fact]
    public void Clamp_TakesANumberThatIsNotANumberBackToTheDefault()
    {
        Assert.Equal(TextMarkMetrics.DefaultFontSize, TextMarkMetrics.Clamp(double.NaN));
        Assert.Equal(TextMarkMetrics.DefaultFontSize, TextMarkMetrics.Clamp(double.PositiveInfinity));
    }

    [Fact]
    public void Fit_PutsTheSecondPointRightOfAndBelowTheAnchorWithoutMovingIt()
    {
        var mark = new AnnotationItem { Kind = EditorTool.Text, Points = [new Point(40, 60)], Text = "Привет", FontSize = 24 };

        TextMarkMetrics.Fit(mark);

        Assert.Equal(new Point(40, 60), mark.Points[0]);
        Assert.Equal(2, mark.Points.Count);
        Assert.True(mark.Points[1].X > mark.Points[0].X);
        Assert.True(mark.Points[1].Y > mark.Points[0].Y);
    }

    [Fact]
    public void Fit_FollowsTheSizeOfTheLettersAndTheirNumber()
    {
        var small = new AnnotationItem { Kind = EditorTool.Text, Points = [new Point(0, 0)], Text = "Привет", FontSize = 12 };
        var large = new AnnotationItem { Kind = EditorTool.Text, Points = [new Point(0, 0)], Text = "Привет", FontSize = 48 };
        var longer = new AnnotationItem { Kind = EditorTool.Text, Points = [new Point(0, 0)], Text = "Привет, мир", FontSize = 12 };

        TextMarkMetrics.Fit(small);
        TextMarkMetrics.Fit(large);
        TextMarkMetrics.Fit(longer);

        Assert.True(large.Points[1].Y > small.Points[1].Y);
        Assert.True(longer.Points[1].X > small.Points[1].X);
    }

    // Every field of a mark has to survive a copy, and this one is the newest of them: a field the
    // clone forgets is a field that disappears on the first undo.
    [Fact]
    public void A_copy_of_a_mark_keeps_the_pattern_of_its_stroke()
    {
        var dashed = new AnnotationItem
        {
            Kind = EditorTool.Rectangle,
            Points = [new Point(10, 10), new Point(80, 40)],
            LineStyle = SnapBrief.Core.Models.AnnotationLineStyle.Dashed
        };

        Assert.Equal(SnapBrief.Core.Models.AnnotationLineStyle.Dashed, dashed.Clone().LineStyle);
        // And the way to the session and back is the same two steps the editor takes.
        var core = dashed.ToCore(800, 600);
        Assert.Equal(SnapBrief.Core.Models.AnnotationLineStyle.Dashed, core.LineStyle);
        Assert.Equal(SnapBrief.Core.Models.AnnotationLineStyle.Dashed, AnnotationItem.FromCore(core, 800, 600).LineStyle);
    }

    [Fact]
    public void Fit_LeavesAMarkThatIsNotACaptionAlone()
    {
        var frame = new AnnotationItem { Kind = EditorTool.Rectangle, Points = [new Point(10, 10), new Point(80, 40)] };

        TextMarkMetrics.Fit(frame);

        Assert.Equal(new Point(80, 40), frame.Points[1]);
    }
}

File: macos/Tests/SnapikCoreTests/TextMarkMetricsTests.swift
// Port of `tests/Snapik.App.Imaging.Tests/TextMarkMetricsTests.cs`, SPEC-DELTA-3 §4, §6.
//
// The numbers are not the Windows ones: the family is the system one here and `Segoe UI Variable
// Text` there. What the facts hold on to is the shape of the box — wider than tall for a word,
// never shorter than the letters, growing with the size and with the number of them — and that is
// the same on both sides.
import XCTest

@testable import SnapikCore

final class TextMarkMetricsTests: XCTestCase {
    func test_Measure_gives_a_wide_box_for_a_word_and_at_least_the_height_of_the_letters() {
        let size = TextMarkMetrics.measure("Привет", fontSize: 32)

        XCTAssertTrue(
            size.width > size.height,
            "A word must be wider than it is tall, it measured \(size.width)x\(size.height).")
        XCTAssertTrue(size.height >= 32)
    }

    func test_Measure_reads_cyrillic_and_latin_alike() {
        let cyrillic = TextMarkMetrics.measure("Привет", fontSize: 20)
        let latin = TextMarkMetrics.measure("Hello", fontSize: 20)

        XCTAssertTrue(cyrillic.width > 0)
        XCTAssertTrue(latin.width > 0)
        XCTAssertTrue(TextMarkMetrics.measure("Привет, мир", fontSize: 20).width > cyrillic.width)
    }

    func test_Measure_of_nothing_is_still_a_box_with_room_for_a_caret() {
        let size = TextMarkMetrics.measure("", fontSize: 20)

        XCTAssertTrue(size.width > 0)
        XCTAssertTrue(size.height >= 20)
    }

    func test_Measure_grows_with_the_size_of_the_letters() {
        XCTAssertTrue(
            TextMarkMetrics.measure("Привет", fontSize: 48).width
                > TextMarkMetrics.measure("Привет", fontSize: 16).width)
    }

    func test_Clamp_brings_a_size_nobody_can_draw_back_into_range() {
        XCTAssertEqual(TextMarkMetrics.minimumFontSize, TextMarkMetrics.clamp(0))
        XCTAssertEqual(TextMarkMetrics.minimumFontSize, TextMarkMetrics.clamp(4))
        XCTAssertEqual(20, TextMarkMetrics.clamp(20))
        XCTAssertEqual(TextMarkMetrics.maximumFontSize, TextMarkMetrics.clamp(400))
    }

    func test_Clamp_takes_a_number_that_is_not_a_number_back_to_the_default() {
        XCTAssertEqual(TextMarkMetrics.defaultFontSize, TextMarkMetrics.clamp(Double.nan))
        XCTAssertEqual(TextMarkMetrics.defaultFontSize, TextMarkMetrics.clamp(.infinity))
    }

    func test_Every_preset_of_the_size_button_is_a_size_the_clamp_leaves_alone() {
        for preset in TextMarkMetrics.fontSizePresets {
            XCTAssertEqual(preset, TextMarkMetrics.clamp(preset))
        }
    }

    func test_Fit_puts_the_second_point_right_of_and_below_the_anchor_without_moving_it() {
        let anchor = GeometryPoint(40, 60)

        let fitted = TextMarkMetrics.fit(anchor: anchor, text: "Привет", fontSize: 24)

        XCTAssertEqual(GeometryPoint(40, 60), anchor)
        XCTAssertTrue(fitted.x > anchor.x)
        XCTAssertTrue(fitted.y > anchor.y)
    }

    func test_Fit_follows_the_size_of_the_letters_and_their_number() {
        let origin = GeometryPoint(0, 0)

        let small = TextMarkMetrics.fit(anchor: origin, text: "Привет", fontSize: 12)
        let large = TextMarkMetrics.fit(anchor: origin, text: "Привет", fontSize: 48)
        let longer = TextMarkMetrics.fit(anchor: origin, text: "Привет, мир", fontSize: 12)

        XCTAssertTrue(large.y > small.y)
        XCTAssertTrue(longer.x > small.x)
    }
}
